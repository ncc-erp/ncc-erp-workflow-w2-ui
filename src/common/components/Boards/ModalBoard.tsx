import { useEffect, useMemo, useRef } from 'react';

import {
  AlertDialog,
  AlertDialogBody,
  AlertDialogContent,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogOverlay,
  Box,
  Button,
  FormControl,
  FormHelperText,
  FormLabel,
  Input,
  Spinner,
  Text,
  VStack,
} from '@chakra-ui/react';
import { useForm } from 'react-hook-form';
import { TextareaField } from '../TextareaField';
import { ErrorMessage } from '@hookform/error-message';
import { ErrorDisplay } from '../ErrorDisplay';

interface ModalBoardProps {
  isOpen: boolean;
  onClose: () => void;
  onConfirm: (data?: string) => void;
  showReason: boolean;
  showDynamicForm?: boolean;
  dynamicForm?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  setReason: (data: string) => void;
}

interface IDynamicFormProps {
  [key: string]: string;
}

const ModalBoard = (props: ModalBoardProps): JSX.Element => {
  const cancelRef = useRef(null);
  const {
    isDisabled = false,
    isLoading = false,
    isOpen,
    onClose,
    onConfirm,
    showReason = false,
    showDynamicForm = false,
    dynamicForm = '',
    setReason,
  } = props;

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm({
    criteriaMode: 'all',
  });

  useEffect(() => {
    reset();
  }, [reset, showDynamicForm]);

  const dynamicFormParse = useMemo(() => {
    if (showDynamicForm) {
      return JSON.parse(dynamicForm);
    }
    return {};
  }, [dynamicForm, showDynamicForm]);

  const toDisplayName = (inputName: string) => {
    return inputName.replace(/([a-z])([A-Z])/g, '$1 $2');
  };

  const renderFormContent = (data: IDynamicFormProps[] | undefined) => {
    return data?.map(function (element, ind) {
      return (
        <FormControl key={ind}>
          <FormLabel fontSize={16} my={1} fontWeight="normal">
            {toDisplayName(element.name)}
            {element.isRequired ? (
              <FormHelperText my={1} style={{ color: 'red' }} as="span">
                {' '}
                *
              </FormHelperText>
            ) : (
              ''
            )}
          </FormLabel>
          <TextareaField
            {...register(element.name, {
              required: element.isRequired
                ? `${toDisplayName(element.name)} is Required`
                : false,
            })}
          />
          <ErrorMessage
            errors={errors}
            name={element.name}
            render={({ message }) => <ErrorDisplay message={message} />}
          />
        </FormControl>
      );
    });
  };

  const onSubmit = async (data: IDynamicFormProps) => {
    if (showDynamicForm) {
      for (const item of dynamicFormParse) {
        if (Object.prototype.hasOwnProperty.call(data, item.name)) {
          item.data = data[item.name];
        }
      }
      onConfirm(JSON.stringify(dynamicFormParse));
      return;
    }

    onConfirm();
  };

  return (
    <>
      <AlertDialog
        isOpen={isOpen}
        leastDestructiveRef={cancelRef}
        onClose={onClose}
      >
        <AlertDialogOverlay>
          <AlertDialogContent>
            <AlertDialogHeader fontSize="lg" fontWeight="bold" pb={1}>
              Update status
            </AlertDialogHeader>

            <AlertDialogBody>
              <Text
                whiteSpace="nowrap"
                fontSize="md"
                fontWeight="medium"
                pb={0}
              >
                Do you want to update status ?
              </Text>
              {showDynamicForm && (
                <form
                  style={{ width: '100%', margin: '10px 0' }}
                  onSubmit={handleSubmit(onSubmit)}
                >
                  <VStack spacing="14px" alignItems="flex-start">
                    {renderFormContent(dynamicFormParse)}
                  </VStack>
                </form>
              )}
              {showReason && (
                <Box mt={5}>
                  <Text
                    whiteSpace="nowrap"
                    fontSize="md"
                    fontWeight="medium"
                    pb={2}
                    pt={0}
                  >
                    Reason *
                  </Text>
                  <Input
                    type="text"
                    onChange={(e) => setReason(e.target.value)}
                  />
                </Box>
              )}
            </AlertDialogBody>

            <AlertDialogFooter>
              <Button onClick={onClose}>Cancel</Button>
              <Button
                colorScheme="red"
                onClick={handleSubmit(onSubmit)}
                ml={3}
                isDisabled={isDisabled || isLoading}
              >
                {isLoading ? <Spinner /> : 'Confirm'}
              </Button>
            </AlertDialogFooter>
          </AlertDialogContent>
        </AlertDialogOverlay>
      </AlertDialog>
    </>
  );
};

export default ModalBoard;
