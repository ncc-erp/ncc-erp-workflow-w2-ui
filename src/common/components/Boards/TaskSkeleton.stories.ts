import type { Meta, StoryObj } from '@storybook/react';
import TaskSkeleton from './TaskSkeleton';

const meta = {
  title: 'W2/TaskSkeleton',
  component: TaskSkeleton,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof TaskSkeleton>;

export default meta;
type Story = StoryObj<typeof meta>;
export const Primary: Story = {};