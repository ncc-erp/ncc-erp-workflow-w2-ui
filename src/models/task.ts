import {
  FetchNextPageOptions,
  InfiniteData,
  InfiniteQueryObserverResult,
  QueryObserverResult,
  RefetchOptions,
  RefetchQueryFilters,
} from '@tanstack/react-query';
import { ListResult } from './request';

export type TaskResult = ListResult<ITask>;
export type StakeHolderResult = ListResult<IStakeHolder>;

export interface ITask {
  name: string;
  email: string;
  emailTo: string[];
  reason?: string;
  status: number;
  workflowInstanceId: string;
  creationTime: string;
  id: string;
  otherActionSignals?: IOtherActionSignals[];
  description?: string;
  authorName?: string;
}

export interface IStakeHolder {
  name: string;
  email: string;
}

export interface IOtherActionSignals {
  otherActionSignal: string;
  status: number;
}

export interface FilterTasks {
  status?: number;
  workflowDefinitionId?: string;
  maxResultCount: number;
  skipCount: number;
  keySearch?: string;
  dates?: string;
}

export interface IRequestUser {
  email: string;
  name: string;
  branchName: string;
}

export interface IRequest {
  CurrentOffice: string;
  Project?: string;
  Device?: string;
  Reason?: string;
  Dates?: string;
  Equipment?: string;
  DestinationOffice?: string;
  Content?: string;
  StartDate?: string;
  EndDate?: string;
}

export interface IInputRequest {
  Request: Record<string, Record<string, string> | string>;
  RequestUser: IRequestUser;
}

export interface ITaskResult {
  input: IInputRequest;
  tasks: ITask;
  emailTo: Array<string>;
  otherActionSignals: IOtherActionSignals[];
}

export type FetchNextPageFunction = (
  options?: FetchNextPageOptions | undefined
) => Promise<InfiniteQueryObserverResult<ITaskResponse, unknown>>;

export type Refetch = <TPageData>(
  options?: (RefetchOptions & RefetchQueryFilters<TPageData>) | undefined
) => Promise<QueryObserverResult<InfiniteData<ITaskResponse>, unknown>>;

export interface ICountTask {
  countPending: number;
  countApproved: number;
  countRejected: number;
}

export interface ITaskRequest {
  maxResultCount: number;
  skipCount: number;
}

export interface ITaskResponse {
  items: ITask[];
  totalCount: number;
}

export interface IInfinityTaskResponse {
  pageParams: number[];
  pages: ITaskPage[];
}

export interface ITaskPage {
  data: ITaskResponse;
}
