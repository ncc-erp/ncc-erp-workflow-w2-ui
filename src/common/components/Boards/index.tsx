import {
  DragDropContext,
  Draggable,
  DropResult,
  Droppable,
} from '@hello-pangea/dnd';
import { useEffect, useState } from 'react';
import styles from './style.module.scss';
import { useQueryClient } from '@tanstack/react-query';
import { toast } from 'common/components/StandaloneToast';
import { BoardColumnStatus, QueryKeys, TaskStatus } from 'common/constants';
import useBoard from './useBoard';
import { Box, Flex, IconButton, Text, useDisclosure } from '@chakra-ui/react';
import { FetchNextPageFunction, ITask, Refetch, TaskResult } from 'models/task';
import { useApproveTask, useRejectTask } from 'api/apiHooks/taskHooks';
import ModalBoard from './ModalBoard';
import { ETaskStatus } from 'common/enums';
import { useCurrentUser } from 'hooks/useCurrentUser';
import { formatDate } from 'utils/formatDate';
import { getDayAgo } from 'utils/getDayAgo';
import { HiArrowDown } from 'react-icons/hi';
import { TaskDetailModal } from 'features/Tasks/components/TaskDetailModal';

interface ModalDetail {
  isOpen: boolean;
  taskId: string;
}

export interface BoardsProps {
  data: {
    listPending: TaskResult;
    listApproved: TaskResult;
    listRejected: TaskResult;
  };
  fetchNextPagePending: FetchNextPageFunction;
  fetchNextPageApproved: FetchNextPageFunction;
  fetchNextPageRejected: FetchNextPageFunction;
  refetchPending: Refetch;
  refetchApproved: Refetch;
  refetchRejected: Refetch;
  status: number;
}

const initialModalStatus: ModalDetail = {
  isOpen: false,
  taskId: '',
};

const Boards = ({
  data,
  fetchNextPageApproved,
  fetchNextPagePending,
  fetchNextPageRejected,
  refetchApproved,
  refetchRejected,
  status,
}: BoardsProps): JSX.Element => {
  const [modalState, setModalState] = useState(initialModalStatus);
  const [result, setResult] = useState<DropResult>();
  const [isRejected, setIsRejected] = useState<boolean>(false);
  const [reason, setReason] = useState<string>('');
  const [state, setState] = useState<Record<ETaskStatus, ITask[]>>({
    [ETaskStatus.Pending]: [],
    [ETaskStatus.Approved]: [],
    [ETaskStatus.Rejected]: [],
  });
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const { isOpen, onOpen, onClose } = useDisclosure();
  const queryClient = useQueryClient();
  const approveTaskMutation = useApproveTask();
  const rejectTaskMutation = useRejectTask();
  const { reorder, move, getItemStyle, getListStyle } = useBoard();
  const currentUser = useCurrentUser();

  const openModal = (taskId: string) => {
    setModalState({
      ...modalState,
      isOpen: true,
      taskId: taskId,
    });
  };

  const closeModal = () => {
    setModalState({
      ...modalState,
      isOpen: false,
    });
  };

  const handleClose = () => {
    onClose();
    setIsRejected(false);
    setReason('');
  };

  const onDragEnd = async (result: DropResult) => {
    const { destination, source } = result;
    if (!destination) return;
    const sInd = +source.droppableId as ETaskStatus;
    const dInd = +destination.droppableId;
    if (sInd === dInd) {
      const items = reorder(state[sInd], source.index, destination.index);
      const newState = { ...state };
      newState[sInd] = items;

      setState(newState);
      return;
    }
    if (+destination.droppableId === BoardColumnStatus.Rejected) {
      setIsRejected(true);
    }
    onOpen();
    setResult(result);
  };

  const handleDrop = async () => {
    const { source, destination } = result as DropResult;
    if (!destination) return;

    const sInd = +source.droppableId as ETaskStatus;
    const dInd = +destination.droppableId as ETaskStatus;

    const results = move(state[sInd], state[dInd], source, destination);

    try {
      setIsLoading(true);
      switch (Number(destination.droppableId)) {
        case BoardColumnStatus.Approved:
          await approveTaskMutation.mutateAsync(
            state[ETaskStatus.Pending][source.index].id
          );
          queryClient.invalidateQueries({
            queryKey: [QueryKeys.FILTER_TASK],
          });
          state[ETaskStatus.Pending][source.index].status = TaskStatus.Approved;
          refetchApproved();
          toast({ title: 'Approved successfully!', status: 'success' });
          break;
        case BoardColumnStatus.Rejected:
          if (!reason) return;
          await rejectTaskMutation.mutateAsync({
            id: state[ETaskStatus.Pending][source.index].id,
            reason,
          });
          queryClient.invalidateQueries({
            queryKey: [QueryKeys.FILTER_TASK],
          });
          state[ETaskStatus.Pending][source.index].status = TaskStatus.Rejected;
          refetchRejected();
          toast({ title: 'Rejected successfully!', status: 'success' });
          break;
        default:
          break;
      }

      const newState = { ...state };
      newState[sInd] = results[sInd];
      newState[dInd] = results[dInd];

      setState(newState);
      handleClose();
    } catch (error) {
      console.error(error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    setState({
      [ETaskStatus.Pending]: data.listPending.items.filter(
        (x) => x.status === TaskStatus.Pending
      ),
      [ETaskStatus.Approved]: data.listApproved.items.filter(
        (x) => x.status === TaskStatus.Approved
      ),
      [ETaskStatus.Rejected]: data.listRejected.items.filter(
        (x) => x.status === TaskStatus.Rejected
      ),
    });
  }, [data]);

  const showMoreItems = (
    fetchNextPage: FetchNextPageFunction,
    listLength: number,
    totalCount: number
  ) => {
    if (listLength < totalCount) {
      return (
        <Flex w={'100%'} justifyContent={'center'} my={2}>
          <IconButton
            variant="solid"
            aria-label="Call Sage"
            fontSize="20px"
            icon={<HiArrowDown />}
            isRound={true}
            colorScheme="teal"
            onClick={() => fetchNextPage()}
          />
        </Flex>
      );
    }
    return null;
  };

  return (
    <>
      <DragDropContext onDragEnd={onDragEnd}>
        <div className={styles.container}>
          {Object.values(state).map((el, ind) => (
            <Droppable key={ind} droppableId={`${ind}`}>
              {(provided, snapshot) => (
                <div
                  ref={provided.innerRef}
                  style={getListStyle(snapshot.isDraggingOver)}
                  {...provided.droppableProps}
                >
                  <div className={styles.columnLabel}>
                    {Object.keys(BoardColumnStatus)[ind]}
                  </div>

                  <Box className={styles.columnContent}>
                    {el.map((item, index) => {
                      const isDisabled =
                        +item.status !== +TaskStatus.Pending ||
                        item?.email !== currentUser?.email;
                      return (
                        <Draggable
                          key={item.id}
                          draggableId={item.id}
                          index={index}
                          isDragDisabled={isDisabled}
                        >
                          {(provided, snapshot) => (
                            <Box
                              cursor={isDisabled ? 'pointer' : 'grab'}
                              onClick={() => {
                                item.id !== null && openModal(item.id);
                              }}
                              ref={provided.innerRef}
                              {...provided.draggableProps}
                              {...provided.dragHandleProps}
                              style={getItemStyle(
                                snapshot.isDragging,
                                provided.draggableProps.style
                              )}
                            >
                              <div
                                className={`${styles.item} ${
                                  ind === BoardColumnStatus.Pending
                                    ? styles.itemPending
                                    : ind === BoardColumnStatus.Approved
                                    ? styles.itemApproved
                                    : ind === BoardColumnStatus.Rejected
                                    ? styles.itemRejected
                                    : ''
                                }`}
                              >
                                <Flex
                                  justifyContent={'space-between'}
                                  w={'100%'}
                                >
                                  <Text fontWeight={'bold'}>
                                    ID: {item.id.slice(-5).toUpperCase()}
                                  </Text>
                                  {getDayAgo(item?.creationTime)}
                                </Flex>
                                <div className={styles.title}>{item.name}</div>

                                <Flex gap={2}>
                                  <Text>Name:</Text> {item.authorName}
                                </Flex>
                                <Flex gap={2}>
                                  <Text>Assign:</Text> {item.email}
                                </Flex>
                                <div className={styles.stateWrapper}>
                                  <div className={styles.state}>State:</div>
                                  <div className={styles.statusWrapper}>
                                    <div
                                      className={`${styles.status} ${
                                        ind === BoardColumnStatus.Pending
                                          ? styles.statusPending
                                          : ind === BoardColumnStatus.Approved
                                          ? styles.statusApproved
                                          : ind === BoardColumnStatus.Rejected
                                          ? styles.statusRejected
                                          : ''
                                      }`}
                                    />
                                    {Object.keys(BoardColumnStatus)[ind]}
                                  </div>
                                </div>
                                <Flex gap={2}>
                                  <Text>Date:</Text>
                                  {formatDate(new Date(item?.creationTime))}
                                </Flex>
                              </div>
                            </Box>
                          )}
                        </Draggable>
                      );
                    })}

                    {ind === BoardColumnStatus.Pending &&
                      (+status === -1 || +status === TaskStatus.Pending) &&
                      showMoreItems(
                        fetchNextPagePending,
                        data?.listPending?.items?.length,
                        data?.listPending?.totalCount
                      )}
                    {ind === BoardColumnStatus.Approved &&
                      (+status === -1 || +status === TaskStatus.Approved) &&
                      showMoreItems(
                        fetchNextPageApproved,
                        data?.listApproved?.items?.length,
                        data.listApproved?.totalCount
                      )}
                    {ind === BoardColumnStatus.Rejected &&
                      (+status === -1 || +status === TaskStatus.Rejected) &&
                      showMoreItems(
                        fetchNextPageRejected,
                        data?.listRejected?.items?.length,
                        data?.listRejected?.totalCount
                      )}
                  </Box>
                  {provided.placeholder}
                </div>
              )}
            </Droppable>
          ))}
        </div>
      </DragDropContext>

      {modalState.taskId.length > 0 && (
        <TaskDetailModal
          isOpen={modalState.isOpen}
          onClose={closeModal}
          taskId={modalState.taskId}
        />
      )}
      <ModalBoard
        isOpen={isOpen}
        onClose={handleClose}
        onConfirm={handleDrop}
        showReason={isRejected}
        setReason={setReason}
        isDisabled={isRejected && !reason}
        isLoading={isLoading}
      />
    </>
  );
};

export default Boards;
