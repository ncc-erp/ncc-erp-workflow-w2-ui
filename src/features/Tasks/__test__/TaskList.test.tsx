import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ListTask } from 'common/components/Boards/ListTask';
import { FilterTasks } from 'models/task';
import { RecoilRoot } from 'recoil';

jest.mock('../../../api/apiHooks/index', () => ({
  useAxios: jest.fn(),
}));

jest.mock('../../../api/axiosInstant', () => ({
  VITE_API_BASE_URL: '/api',
}));

jest.mock('hooks/useIsAdmin', () => ({
  useIsAdmin: jest.fn().mockReturnValue(true),
}));

jest.mock('common/components/WorkflowModal', () => ({
  VITE_GOOGLE_LOGIN_REDIRECT: 'http://localhost:4200',
}));

jest.mock('hooks/useCurrentUser', () => ({
  useCurrentUser: jest.fn().mockReturnValue({ email: 'bob@example.com' }),
}));

jest.mock('api/apiHooks/taskHooks', () => ({
  useGetTasks: jest.fn().mockReturnValue({
    isLoading: false,
    isRefetching: false,
    refetch: jest.fn(),
    data: {
      totalCount: 1,
      items: [
        {
          id: '3a0e2b21-7143-224b-f17f-6d0ca4f7d274',
          workflowInstanceId: '3a0e2b21-6ea7-959e-fe4c-e5954ead2832',
          workflowDefinitionId: '3a0e256f-a848-065c-fab4-56e8c133918b',
          email: null,
          status: 1,
          name: 'Probationary Request',
          description: 'PM Reviews',
          dynamicActionData:
            '[{"name":"StrengthPoints","type":"RichText","isRequired":true,"data":""},{"name":"WeaknessPoints","type":"RichText","isRequired":true,"data":""}]',
          reason: null,
          creationTime: '2023-10-10T09:39:25.124704Z',
          otherActionSignals: null,
          emailTo: ['trang.danghuyen@ncc.asia', 'long.vodinhhoang@ncc.asia'],
          author: '3a0daa05-f851-d09a-9361-8896cc20637b',
          authorName: 'Bubu chacha',
        },
      ],
    },
  }),
  useActionTask: jest.fn(() => ({
    mutateAsync: jest.fn(),
  })),

  useApproveTask: jest.fn(() => ({
    mutateAsync: jest.fn(),
  })),

  useRejectTask: jest.fn(() => ({
    mutateAsync: jest.fn(),
  })),
}));

test('Request My Requests Page', () => {
  const filter: FilterTasks = {
    maxResultCount: 10,
    skipCount: 0,
  };
  const queryClient: QueryClient = new QueryClient();
  const { container } = render(
    <QueryClientProvider client={queryClient}>
      <RecoilRoot>
        <ListTask filters={filter} openDetailModal={() => jest.fn()} />
      </RecoilRoot>
    </QueryClientProvider>
  );
  expect(container).toMatchSnapshot();
});
