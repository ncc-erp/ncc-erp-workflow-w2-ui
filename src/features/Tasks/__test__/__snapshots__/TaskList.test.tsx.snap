// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request My Requests Page 1`] = `
<div>
  <div
    class="css-79elbk"
  >
    <button
      aria-label="Done"
      class="chakra-button css-1g6czxt"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M909.1 209.3l-56.4 44.1C775.8 155.1 656.2 92 521.9 92 290 92 102.3 279.5 102 511.5 101.7 743.7 289.8 932 521.9 932c181.3 0 335.8-115 394.6-276.1 1.5-4.2-.7-8.9-4.9-10.3l-56.7-19.5a8 8 0 0 0-10.1 4.8c-1.8 5-3.8 10-5.9 14.9-17.3 41-42.1 77.8-73.7 109.4A344.77 344.77 0 0 1 655.9 829c-42.3 17.9-87.4 27-133.8 27-46.5 0-91.5-9.1-133.8-27A341.5 341.5 0 0 1 279 755.2a342.16 342.16 0 0 1-73.7-109.4c-17.9-42.4-27-87.4-27-133.9s9.1-91.5 27-133.9c17.3-41 42.1-77.8 73.7-109.4 31.6-31.6 68.4-56.4 109.3-73.8 42.3-17.9 87.4-27 133.8-27 46.5 0 91.5 9.1 133.8 27a341.5 341.5 0 0 1 109.3 73.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 0 0 3 14.1l175.6 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c-.1-6.6-7.8-10.3-13-6.2z"
        />
      </svg>
    </button>
    <div
      class="css-1t3mkjm"
    >
      <div
        class="tableContent css-kge0eu"
      >
        <table
          class="chakra-table css-zy166a"
        >
          <thead
            class="css-0"
          >
            <tr
              class="css-1y39i0r"
            >
              <th
                class="css-avwfu8"
                colspan="1"
                style="width: auto;"
              >
                <div
                  class="css-alaj7i"
                >
                  <div
                    class="css-n085mf"
                  >
                    ID
                  </div>
                </div>
              </th>
              <th
                class="css-avwfu8"
                colspan="1"
                style="width: auto;"
              >
                <div
                  class="css-alaj7i"
                >
                  <div
                    class="css-xi606m"
                  >
                    Title
                  </div>
                </div>
              </th>
              <th
                class="css-avwfu8"
                colspan="1"
                style="width: auto;"
              >
                <div
                  class="css-alaj7i"
                >
                  Request user
                </div>
              </th>
              <th
                class="css-avwfu8"
                colspan="1"
                style="width: auto;"
              >
                <div
                  class="css-alaj7i"
                >
                  Current State
                </div>
              </th>
              <th
                class="css-avwfu8"
                colspan="1"
                style="width: auto;"
              >
                <div
                  class="css-alaj7i"
                >
                  Assigned To
                </div>
              </th>
              <th
                class="css-avwfu8"
                colspan="1"
                style="width: auto;"
              >
                <div
                  class="css-alaj7i"
                >
                  State
                </div>
              </th>
              <th
                class="css-avwfu8"
                colspan="1"
                style="width: auto;"
              >
                <div
                  class="css-alaj7i"
                >
                  Created At
                </div>
              </th>
              <th
                class="css-avwfu8"
                colspan="1"
                style="width: auto;"
              >
                <div
                  class="css-alaj7i"
                >
                  <div
                    class="css-mgpx6q"
                  >
                    Actions
                  </div>
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class="css-0"
          >
            <tr
              class="css-1bgkdca"
            >
              <td
                class="css-xk16pm"
              >
                <div
                  class="css-gmuwbf"
                >
                  7D274
                </div>
              </td>
              <td
                class="css-xk16pm"
              >
                <div
                  class="css-0"
                  style="display: flex; align-items: start; flex-direction: column; gap: 5px; min-width: 400px;"
                >
                  <div
                    class="css-0"
                    style="flex: 1; display: flex; width: 100%;"
                  >
                    <p
                      class="chakra-text css-e3cc90"
                      data-tooltip-id=""
                      style="overflow: hidden; text-overflow: ellipsis; font-weight: 600; display: inline-block; flex: 1; width: 0px; white-space: nowrap; margin: 0px;"
                    />
                  </div>
                  <div
                    class="titleBoard css-0"
                    style="background-color: rgb(170, 187, 204);"
                  >
                    <div
                      style="overflow: hidden; text-overflow: ellipsis; display: -webkit-box;"
                    >
                      Probationary Request
                    </div>
                  </div>
                </div>
              </td>
              <td
                class="css-xk16pm"
              >
                <div
                  class="css-0"
                >
                  Bubu chacha
                </div>
              </td>
              <td
                class="css-xk16pm"
              >
                PM Reviews
              </td>
              <td
                class="css-xk16pm"
              >
                trang.danghuyen, long.vodinhhoang
              </td>
              <td
                class="css-xk16pm"
              >
                <div
                  class="css-k008qs"
                >
                  <div
                    class="badge statusApproved"
                  >
                    Approved
                  </div>
                </div>
              </td>
              <td
                class="css-xk16pm"
              >
                10-10-2023 16:39
              </td>
              <td
                class="css-xk16pm"
              >
                <div
                  class="css-h607p5"
                >
                  <div
                    class="tableActionLoading"
                  />
                  <button
                    aria-controls="menu-list-:rg:"
                    aria-expanded="false"
                    aria-haspopup="menu"
                    aria-label=""
                    class="chakra-button chakra-menu__menu-button css-p7mkev"
                    id="menu-button-:rg:"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="chakra-icon css-bzk4kq"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.33409 4.54491C6.3494 3.63637 7.55145 2.9322 8.87555 2.49707C9.60856 3.4128 10.7358 3.99928 12 3.99928C13.2642 3.99928 14.3914 3.4128 15.1245 2.49707C16.4486 2.9322 17.6506 3.63637 18.6659 4.54491C18.2405 5.637 18.2966 6.90531 18.9282 7.99928C19.5602 9.09388 20.6314 9.77679 21.7906 9.95392C21.9279 10.6142 22 11.2983 22 11.9993C22 12.7002 21.9279 13.3844 21.7906 14.0446C20.6314 14.2218 19.5602 14.9047 18.9282 15.9993C18.2966 17.0932 18.2405 18.3616 18.6659 19.4536C17.6506 20.3622 16.4486 21.0664 15.1245 21.5015C14.3914 20.5858 13.2642 19.9993 12 19.9993C10.7358 19.9993 9.60856 20.5858 8.87555 21.5015C7.55145 21.0664 6.3494 20.3622 5.33409 19.4536C5.75952 18.3616 5.7034 17.0932 5.0718 15.9993C4.43983 14.9047 3.36862 14.2218 2.20935 14.0446C2.07212 13.3844 2 12.7002 2 11.9993C2 11.2983 2.07212 10.6142 2.20935 9.95392C3.36862 9.77679 4.43983 9.09388 5.0718 7.99928C5.7034 6.90531 5.75952 5.637 5.33409 4.54491ZM13.5 14.5974C14.9349 13.7689 15.4265 11.9342 14.5981 10.4993C13.7696 9.0644 11.9349 8.57277 10.5 9.4012C9.06512 10.2296 8.5735 12.0644 9.40192 13.4993C10.2304 14.9342 12.0651 15.4258 13.5 14.5974Z"
                      />
                    </svg>
                  </button>
                  <div
                    class="css-0"
                    style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
                  >
                    <div
                      aria-orientation="vertical"
                      class="chakra-menu__menu-list css-2vme5"
                      id="menu-list-:rg:"
                      role="menu"
                      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
                      tabindex="-1"
                    >
                      <button
                        class="chakra-menu__menuitem css-z08cet"
                        data-index="0"
                        id="menu-list-:rg:-menuitem-:rh:"
                        role="menuitem"
                        tabindex="-1"
                        type="button"
                      >
                        <svg
                          class="chakra-icon css-1vzdt75"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M1.18164 12C2.12215 6.87976 6.60812 3 12.0003 3C17.3924 3 21.8784 6.87976 22.8189 12C21.8784 17.1202 17.3924 21 12.0003 21C6.60812 21 2.12215 17.1202 1.18164 12ZM12.0003 17C14.7617 17 17.0003 14.7614 17.0003 12C17.0003 9.23858 14.7617 7 12.0003 7C9.23884 7 7.00026 9.23858 7.00026 12C7.00026 14.7614 9.23884 17 12.0003 17ZM12.0003 15C10.3434 15 9.00026 13.6569 9.00026 12C9.00026 10.3431 10.3434 9 12.0003 9C13.6571 9 15.0003 10.3431 15.0003 12C15.0003 13.6569 13.6571 15 12.0003 15Z"
                          />
                        </svg>
                        View
                      </button>
                      <button
                        class="chakra-menu__menuitem css-z08cet"
                        data-index="1"
                        id="menu-list-:rg:-menuitem-:ri:"
                        role="menuitem"
                        tabindex="-1"
                        type="button"
                      >
                        <svg
                          class="chakra-icon css-zvw1k8"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M2 5L9 2L15 5L21.303 2.2987C21.5569 2.18992 21.8508 2.30749 21.9596 2.56131C21.9862 2.62355 22 2.69056 22 2.75827V19L15 22L9 19L2.69696 21.7013C2.44314 21.8101 2.14921 21.6925 2.04043 21.4387C2.01375 21.3765 2 21.3094 2 21.2417V5Z"
                          />
                        </svg>
                        Workflow
                      </button>
                    </div>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="chakra-stack css-m4cjlb"
    >
      <div
        class="chakra-stack css-130amzp"
      >
        <div
          class="chakra-stack css-1hohgv6"
        >
          <p
            class="chakra-text css-1ok354g"
          >
            Rows per page
          </p>
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <div
              class="chakra-select__wrapper css-bo0or5"
            >
              <select
                class="chakra-select css-1y1ernp"
                id="field-:r9:"
              >
                <option
                  value="10"
                >
                  10
                </option>
                <option
                  value="25"
                >
                  25
                </option>
                <option
                  value="50"
                >
                  50
                </option>
                <option
                  value="100"
                >
                  100
                </option>
              </select>
              <div
                class="chakra-select__icon-wrapper css-10n8c1x"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-icon chakra-select__icon css-13otjrl"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  role="presentation"
                  stroke="currentColor"
                  stroke-width="0"
                  style="color: currentColor; width: 1em; height: 1em;"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 14L8 10H16L12 14Z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-iz7l97"
        />
        <p
          class="chakra-text css-3z0v2v"
        >
          showing 1 to 1 of 1 entries
        </p>
      </div>
    </div>
  </div>
</div>
`;
